### PR TITLE
feat(engine): @lwc/node-reactions

### PR DESCRIPTION
**Pull request by ravijayaramappa**
## Details
Library to react to dom elements connecting and disconnecting from the document
RFC: https://github.com/salesforce/lwc-rfcs/pull/11

# POC
https://github.com/salesforce/lwc/pull/1431

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

If yes, please describe the impact and migration path for existing applications.

## The PR fulfills these requirements:
* Have tests for the proposed changes been added? ✅
* Have you followed [these instructions](../CONTRIBUTING.md#-commit-message-conventions) to clearly describe the issue being fixed or feature enhanced? ✅

### CODE REVIEW

File: packages/@lwc/node-reactions/src/global/init.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+const InitializationSlot = '$$node-reactions-initialized$$';
+let reactWhenConnectedCached: (elm: Element, callback: ReactionCallback) => void;
+let reactWhenDisconnectedCached: (elm: Element, callback: ReactionCallback) => void;
+
+initialize();
+/**
+ * Set an internal field to detect initialization
+ */
+export function initialize() {
```
Comment (pmdartus) [inline packages/@lwc/node-reactions/src/global/init.ts]: This function is not used outside the module and should not be exported.

File: packages/@lwc/node-reactions/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import './global/init';
```
Comment (pmdartus) [inline packages/@lwc/node-reactions/src/index.ts]: IMO, it would make sense to consolidate the `index.ts`, `api.ts` and `init.ts` in a single `index.ts` file. It feels wrong that the `api.ts` imports `init.ts`.

File: packages/@lwc/node-reactions/src/core/reactions.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        // TODO: Can we substitute with an instanceof check?
+        // duck-typing to detect if node is an element or a document fragment, instead of the expensive instanceOf
+        'childElementCount' in elmOrDocFrag &&
+        ((elmOrDocFrag as Element | DocumentFragment).childElementCount > 0 ||
+            isRegisteredNode(elmOrDocFrag))
+    );
+}
+
+export function isQualifyingHost(node: Node): boolean {
+    return node != null && isRegisteredNode(node);
```
Comment (pmdartus) [inline packages/@lwc/node-reactions/src/core/reactions.ts]: If this function accepts a `Node`, `node != null` will always be true. This check feels unnecessary.

File: packages/@lwc/node-reactions/src/core/traverse.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * Process nodes in a NodeList of marked elements
+ */
+function queueReactionsForNodeList(
+    nodeList: NodeList,
+    reactionTypes: QualifyingReactionTypes,
+    reactionQueue: ReactionRecord[]
+) {
+    forEach.call(nodeList, node => {
```
Comment (pmdartus) [inline packages/@lwc/node-reactions/src/core/traverse.ts]: Will be faster to iterate using a simple for loop than a `forEach`

File: packages/@lwc/node-reactions/src/core/traverse.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        queueReactionsForSingleElement(node, reactionTypes, reactionQueue);
+        // If node has a shadow tree, process its shadow tree
+        queueReactionsForShadowRoot((node as Element).shadowRoot!, reactionTypes, reactionQueue);
```
Comment (pmdartus) [inline packages/@lwc/node-reactions/src/core/traverse.ts]: I don't see how you can ensure that the `node` has an associated `ShadowRoot`.

File: packages/@lwc/node-reactions/src/core/traverse.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    }
+    if (reactionTypes & 1) {
+        const reactionRecords = getConnectedRecordsForElement(elm);
+        if (!isUndefined(reactionRecords)) {
+            queueReactionRecord(reactionQueue, reactionRecords);
+        }
+    }
+}
+
+const ShadowRootQuerySelectorAll = ShadowRoot.prototype.querySelectorAll;
```
Comment (pmdartus) [inline packages/@lwc/node-reactions/src/core/traverse.ts]: `ShadowRoot` doesn't have a `querySelectorAll` attached from its prototype. It is coming from `DocumentFragment`.

File: packages/@lwc/node-reactions/src/core/reactions.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+ * Is a given node qualified for reactions?
+ * Conditions to satisfy
+ * 1. The root, has to be an element or document fragment
+ * 2. The root, has to have children or is a registered element
+ *  2a. Assumption here is that all custom elements are registered
+ */
+export function isQualifyingElement(elmOrDocFrag: Node): boolean {
+    return (
+        elmOrDocFrag != null &&
+        // TODO: Can we substitute with an instanceof check?
```
Comment (pmdartus) [inline packages/@lwc/node-reactions/src/core/reactions.ts]: Yes, why not?

File: packages/integration-karma/test/component/lifecycle-callbacks/index.spec.js [hunk 8]
```diff
@@ -130,3 +131,54 @@ it('should call children component lifecycle hooks when a public property change
         elm.childVisible = false;
     });
 });
+
+xit('should call parent and children lifecycle hooks in correct order when parent reconnected', () => {
```
Comment (pmdartus) [inline packages/integration-karma/test/component/lifecycle-callbacks/index.spec.js]:
```suggestion
it('should call parent and children lifecycle hooks in correct order when parent reconnected', () => {
```

File: packages/integration-karma/test/component/lifecycle-callbacks/index.spec.js [hunk 9]
```diff
@@ -130,3 +131,54 @@ it('should call children component lifecycle hooks when a public property change
...
+        document.body.removeChild(elm);
+
+        expect(window.timingBuffer).toEqual([
+            'slotted-parent:disconnectedCallback',
+            'accepting-slot:disconnectedCallback',
+            'child:disconnectedCallback',
+        ]);
+    });
+    // TODO: In native shadow mode, accepting-slot:renderedCallback is invoked before child:constructor
+    xit('appending parent to the DOM', () => {
```
Comment (pmdartus) [inline packages/integration-karma/test/component/lifecycle-callbacks/index.spec.js]:
```suggestion
    it('appending parent to the DOM', () => {
```

File: packages/@lwc/engine/src/framework/vm.ts [hunk 10]
```diff
@@ -442,70 +416,32 @@ function runDisconnectedCallback(vm: VM) {
...
-                resetComponentStateWhenRemoved(getCustomElementVM(vnode.elm as HTMLElement));
+    const { children, cmpRoot } = vm;
+    for (let i = 0, len = children.length; i < len; i += 1) {
+        const vnode = children[i];
+        if (!isNull(vnode)) {
+            const { elm } = vnode;
+            if (!isUndefined(elm)) {
+                // the concern here is that this routine might
+                // throw and we have one flow that does not
+                // have protection (recovering from error boundary)
```
Comment (ekashida) [inline packages/@lwc/engine/src/framework/vm.ts]: Is this saying that the only time we expect this routine to throw is when we're recovering from an error boundary?

File: packages/@lwc/node-reactions/jest.config.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+module.exports = {
+    ...BASE_CONFIG,
+
+    displayName: 'lwc-node-reactions',
+
+    roots: ['<rootDir>/src'],
+
+    // Override global threshold for the package. As we increase the test coverage we should increase
+    // the threshold as well.
```
Comment (ekashida) [inline packages/@lwc/node-reactions/jest.config.js]: Is this just a copy/paste comment or is the coverage for this package actually low?

File: packages/@lwc/node-reactions/src/types.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+export type ReactionType = 1 /* connected */ | 2 /* disconnected */;
```
Comment (ekashida) [inline packages/@lwc/node-reactions/src/types.ts]: Any reason why we didn't use [enums](https://www.typescriptlang.org/docs/handbook/enums.html) here?
```
export enum ReactionType {
  Connected,
  Disconnected,
  None,
}
```

File: packages/@lwc/node-reactions/src/dom-patching/node.ts [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+            writable: true,
+            enumerable: true,
+            configurable: true,
+            value: function(this: Node, aChild: Node) {
+                // Performance optimization: Only check if the node being added is a registered node.
+                // We made this decision because appendChild is on the critical path and cannot
+                // subject every node to the expensive check that isQualifyingElement() performs
+                if (!isQualifyingHost(aChild)) {
+                    // invoke with apply to preserve native behavior of missing arguments, invalid types etc
+                    return appendChild.call(this, aChild);
```
Comment (ekashida) [inline packages/@lwc/node-reactions/src/dom-patching/node.ts]: Not sure if we're not doing the following for performance reasons. If so, we should update the comment.
```suggestion
                    return appendChild.apply(this, ArraySlice.call(arguments) as [Node]);
```

File: packages/@lwc/engine/src/framework/stylesheet.ts [hunk 14]
```diff
@@ -83,18 +80,40 @@ export function resetStyleAttributes(vm: VM): void {
...
+    const { context, elm, cmpRoot } = vm;
+    const { hostAttribute, shadowAttribute } = stylesheetTokens;
     // Remove the style attribute currently applied to the host element.
     setAttribute.call(elm, hostAttribute, '');
 
     context.hostAttribute = hostAttribute;
     context.shadowAttribute = shadowAttribute;
+
+    // Caching style vnode so it can be removed when needed
+    const styleElm = (context.styleVNode = evaluateCSS(
```
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/engine/src/framework/stylesheet.ts]: @caridy Do you concur? 
The property name was `styleNode` even before this PR. Any objections to rename this to `styleElement`?

File: packages/@lwc/engine/src/framework/vm.ts [hunk 15]
```diff
@@ -177,15 +165,17 @@ function resetComponentStateWhenRemoved(vm: VM) {
...
-        runDisconnectedCallback(vm);
-        // Spec: https://dom.spec.whatwg.org/#concept-node-remove (step 14-15)
-        runShadowChildNodesDisconnectedCallback(vm);
-        runLightChildNodesDisconnectedCallback(vm);
+    }
+    if (isFalse(vm.isDirty)) {
+        // this guarantees that if the component is reused/reinserted,
+        // it will be re-rendered because we are disconnecting the reactivity
+        // linking, so mutations are not automatically reflected on the state
+        // of disconnected components.
```
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/engine/src/framework/vm.ts]: Going through the code, I think it means
```        
// this guarantees that if the component is reused/reinserted, it will be re-rendered.
// it should be re-rendered because we are disconnecting the reactivity
// linking, so mutations are not automatically reflected on the state
// of disconnected components.
```
@caridy can confirm.

Note: This piece of code was moved here from runDisconnectedCallback()

File: packages/@lwc/engine/src/framework/vm.ts (hunk 10 above)
Comment (ravijayaramappa) [reply packages/@lwc/engine/src/framework/vm.ts]: No, that is not what it says. 
It says that we are recovering from an error here and trying to clean up things. If removeChild() throws an error, its fine to ignore the error. For example, we were trying to clean up a vnode that was created but was never inserted to the shadow root, removeChild() will throw. And it is safe to ignore such an error.

File: packages/@lwc/node-reactions/jest.config.js (hunk 11 above)
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/jest.config.js]: It is from Caridy's initial PR for node-reactions https://github.com/salesforce/lwc/pull/1173/files#diff-9a5b37344286d058fc3a43335eeed323. Updated to just use the settings from the base repo.

File: packages/@lwc/node-reactions/src/core/reactions.ts (hunk 7 above)
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/core/reactions.ts]: For perf reasons:
Screen Shot 2019-10-08 at 1 38 09 PM
https://jsperf.com/duck-typing-vs-instanceof-lwc

File: packages/@lwc/node-reactions/src/core/reactions.ts (hunk 3 above)
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/core/reactions.ts]: The argument can be undefined or null too. Will update the type specification.

File: packages/@lwc/node-reactions/src/core/traverse.ts (hunk 6 above)
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/core/traverse.ts]: In synthetic-shadow mode, the querySelectorAll is resolved on the ShadowRoot's prototype
https://github.com/salesforce/lwc/blob/master/packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts#L561-L590

File: packages/@lwc/node-reactions/src/types.ts [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+
+export type ReactionCallback = (this: Element, reactionType: ReactionType) => void;
```
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/types.ts]: I am changing this to pass the element as the first argument instead of the context.
The original motivation was to model it similar to addEventlistener(). For a listener, the listener is invoked with the element as the context.

File: packages/@lwc/node-reactions/scripts/rollup/rollup.config.js [hunk 17]
```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+const path = require('path');
+const typescript = require('rollup-plugin-typescript');
+
+const { version } = require('../../package.json');
```
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/scripts/rollup/rollup.config.js]: https://github.com/salesforce/lwc/issues/1558

File: packages/@lwc/node-reactions/src/global/README.md [hunk 18]
```diff
@@ -0,0 +1,2 @@
+Note:
+Modules in this folder contain stateful information. This is important when working with unit test environments. Jest for example shares DOM prototypes across test suites. Care must be taken to avoid repeatedly patching prototypes.
```
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/global/README.md]: The readme is not applicable anymore. We were trying to use a Symbol to cache the init routine on the DocumentPrototype. That was causing some issues with JEST. We have switched to a string field since.

[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/global/README.md]: Agreed. This is a common problem. May be we can reset the prototypes at the root Jest config. Resetting at the end of each test suite will become tedious and easy to forget.

File: packages/@lwc/node-reactions/src/api.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+import { isUndefined, assert } from '@lwc/shared';
+import { ReactionCallback } from './types';
+import { reactWhenConnected, reactWhenDisconnected } from './global/init';
+
+/**
+ * Redirecting after dev mode assertion.
+ * In prod mode, the browser should be able to optimize this
+ */
+
+function reactWhenConnectedRedirect(elm: Element, callback: ReactionCallback) {
```
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/api.ts]: We initially started out with any Node in the POC stage. The performance regression was becoming unacceptable for LWC usage. So we started narrowing down the scope to just handling custom elements as a first version of the library. May be in the future with perf improvements, this library will be able to handle any type of node. 🤞

File: packages/@lwc/node-reactions/src/dom-patching/node.ts [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+
+import { getOwnPropertyDescriptor, defineProperties } from '@lwc/shared';
+import { ReactionRecord, QualifyingReactionTypes } from '../types';
+import queueReactionsForSubtree from '../core/traverse';
+import { flushQueue } from '../core/reaction-queue';
+import { isQualifyingElement, isQualifyingHost, marker } from '../core/reactions';
+
+export default function() {
+    // caching few DOM APIs
+    const { appendChild, insertBefore, removeChild, replaceChild } = Node.prototype;
```
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/dom-patching/node.ts]: It is no reused outside of the function body. So kept them as local variables instead of pulling them out.
Also, we need to get the patched version of the APIs, so dereferencing them just before patching them would guarantee that we will always get the patched versions of the APIs(synthetic-shadow).

File: packages/@lwc/node-reactions/src/index.ts (hunk 2 above)
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/index.ts]: init.ts has the initialization routine. So didn't want that to end up in the main file of the module.
api.ts is basically a redirection with some dev mode asserts. Also didn't want that to be the first thing somebody sees in index.ts. I could rename it if you suggest a better name.
index.ts is really the entry point into the library.

File: packages/@lwc/node-reactions/src/core/traverse.ts (hunk 5 above)
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/core/traverse.ts]: So the NodeList is always the result of querying for the marked nodes, we can be sure it always results a NodeListOf<Element>. Will update the type definition for the arguments.

File: packages/@lwc/node-reactions/src/dom-patching/node.ts (hunk 13 above)
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/dom-patching/node.ts]: Yeah, we are not doing that for performance reasons. It caused a 5% perf regression.

Man, you caught the one place that I forgot to remove the comment :)

File: packages/@lwc/node-reactions/src/types.ts (hunk 12 above)
Comment (ravijayaramappa) [reply packages/@lwc/node-reactions/src/types.ts]: We did for perf. If we used enum, it would add a `.` property lookup to the execution path. In the original PR, I started out with an enum. After discussing with Caridy, we moved to this specific flags. So we sacrificed code readability for perf.

File: packages/@lwc/engine/src/__tests__/events.spec.ts [hunk 21]
```diff
@@ -182,8 +182,8 @@ describe('events', () => {
             divWithClickHandler.addEventListener('click', function(event) {
                 target = event.target;
             });
-            divWithClickHandler.appendChild(elm);
             document.body.appendChild(divWithClickHandler);
+            divWithClickHandler.appendChild(elm);
```
[the remark this answers is not in the document]
Comment (ravijayaramappa) [reply packages/@lwc/engine/src/__tests__/events.spec.ts]: It is because the custom element will only be hydrated when it is attached to a connected parent. 

tl;dr;
This was an optimization in node-reaction for appendChild() and insertBefore(). The lib will only trigger the connected callback if the appended child is a qualifying node and the parent is connected.

File: packages/@lwc/node-reactions/src/core/reactions.ts (hunk 7 above)
Comment (pmdartus) [reply packages/@lwc/node-reactions/src/core/reactions.ts]: Extracting this check into its own specific method and adding a comment with a link to the performance benchmark would be great.

File: packages/@lwc/node-reactions/src/types.ts (hunk 12 above)
Comment (pmdartus) [reply packages/@lwc/node-reactions/src/types.ts]: I agree it would be great to avoid the performance penalty. What about exporting constants instead and using those constants instead of number literals?

```js
export const REACTION_TYPE_CONNECTED = 1;
export const REACTION_TYPE_DISCONNECTED = 2;
```

Terser does constant folding and will inline the references automatically for you.

File: packages/@lwc/node-reactions/src/index.ts (hunk 2 above)
Comment (pmdartus) [reply packages/@lwc/node-reactions/src/index.ts]: The strange part to me is that `api.ts` imports `init.ts`. If anything the init file should put the library in a consumable state but should expose properties that other modules consume.

**Comment by ravijayaramappa** [comment]
Rebased and created a new PR https://github.com/salesforce/lwc/pull/1747
